src/core/server/elasticsearch/elasticsearch_service.ts

```diff
@@ -57,6 +61,16 @@ export class ElasticsearchService
       .pipe(map((rawConfig) => new ElasticsearchConfig(rawConfig)));
   }
 
+  public async preboot(): Promise<InternalElasticsearchServicePreboot> {
+    this.log.debug('Prebooting elasticsearch service');
+
+    const config = await this.config$.pipe(first()).toPromise();
```

Comment:
__note:__ `userSetup` preboot plugin will need to know current ES configuration, but we're not sure yet what part of the configuration we need exactly. Once we know that, we can reduce the amount of the config information we expose if we want to.

Reply:
> 1. Why exposing the config instead of an observable?

The main reason was to have the same config as we use in `createClient` under the hood (since we didn't rely on `Observable` there). I can definitely switch to observable here, but will need to refactor existing `createClient` to work with the latest `config` as well. Would that be better?

>  2. Even if you're gonna need the config for the feature, I wonder if we shouldn't put that thing under a legacy property, as it's done for the other contracts. After all, raw ES config access is still something we would really like to avoid, so...

I think it makes to introduce `legacy` only when we have a better path forward. It made sense for the contract we provided to the `standard` plugins as they don't really need access to that kind of information, but it feels reasonable to me to give a little bit more power and knowledge to the "service" `preboot` plugins that have enough "privileges" to configure ES connection themselves. Just my 2 cents.

Reply:
> Are you going to recreate ES client manually for every config update during preboot stage?

Yeah, I guess we don't have other choice than to reconfigure existing client if we need to apply new configuration and see if it works?

Reply:
Also let me know, if you think you need more time to figure out the best solution here. In this case I can stop exposing `config` in this PR and get back to this question when we introduce `userSetup` plugin.

Just for the context, at this point, we know that we'll need a way to just figure out if connection to ES is fully configured. It should include at least four checks:

* If username/password are set
OR
* If `kibana_system` user service account token is set
OR
* If Kibana "client" certificate is configured (in case user wants PKI authentication between Kibana and ES)
OR
* If `elasticsearch.certificateAuthorities` configured

I didn't want to keep these checks in Core and hence exposed the `config`, so that `userSetup` plugin can implement that logic instead. We cannot just rely on the connection attempt outcome as it can easily be because of invalid configuration that user provided, in this case we will activate interactive setup mode.

Reply:
I'm okay to expose `RecursiveReadonly<ElasticsearchConfig>` config to `preboot` plugins. 
Super nit: What if we expose the required sub-set of fields? It would simplify our life in the future whenever we modify `ElasticsearchConfig` interface.

Reply:
> Super nit: What if we expose the required sub-set of fields? It would simplify our life in the future whenever we modify ElasticsearchConfig interface.

Yeah, I can do this as well, for this PR I can limit that to username/password/certificateAuthorities and if we figure out that we need more (or less) in https://github.com/elastic/kibana/pull/105222, we'll adjust this interface. Changes will be within single release, so there shouldn't be any BWC concerns (ignoring alpha1).

Reply:
Done in https://github.com/elastic/kibana/pull/103636/commits/ded628fdd3893696bbdee8c503d1597062cfe3b9